File: data_structures_by_c++/stack/stack_using_singly_linked_list.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int val;
    Node *next;
    Node(int val)
    {
        this->val = val;
        this->next = NULL;
    }
};

class my_stack
{
public:
    Node *head = NULL;
    int sz = 0;

    void push(int val) // O(1)
    {
        sz++;
        Node *newnode = new Node(val);
        newnode->next = head;
        head = newnode;
    }

    void pop() // O(1)
    {
        if (head == NULL)
            return;
        sz--;
        Node *deletenode = head;
        head = head->next;
        delete deletenode;
    }

    int top() // O(1)
    {
        return head->val;
    }

    int size() // O(1)
    {
        return sz;
    }

    bool empty() // O(1)
    {
        return head == NULL;
    }
};

int main()
{
    my_stack st;

    st.push(10);
    st.push(20);
    st.push(30);
    st.push(40);

    cout << st.top() << endl;

    while (!st.empty())
    {
        cout << st.top() << " ";
        st.pop();
    }
}

File: data_structures_by_python/stack/stack_using_singly_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class my_stack:
    def __init__(self):
        self.head = None
        self.sz = 0

    def push(self, val):  # O(1)
        self.sz += 1
        newnode = Node(val)
        newnode.next = self.head
        self.head = newnode

    def pop(self):  # O(1)
        if self.head is None:
            return
        self.sz -= 1
        self.head = self.head.next

    def top(self):  # O(1)
        return self.head.val

    def size(self):  # O(1)
        return self.sz

    def empty(self):  # O(1)
        return self.head is None


st = my_stack()

st.push(10)
st.push(20)
st.push(30)
st.push(40)

print(st.top())

while not st.empty():
    print(st.top(), end=" ")
    st.pop()
